VDPO (#2148)
huggingface/blog: dpo_vlm.md

        chosen = processor.apply_chat_template(chosen, tokenize=False)
        rejected = processor.apply_chat_template(rejected, tokenize=False)
        # Resize the image to ensure it fits within the maximum allowable
        # size of the processor to prevent OOM errors.
        max_size = processor.image_processor.size["longest_edge"] // 2
        example["image"].thumbnail((max_size, max_size))
        return {"images": [example["image"]], "prompt": prompt, "chosen": chosen, "rejected": rejected}

    # Apply the formatting function to the dataset
    dataset = dataset.map(format, remove_columns=dataset.column_names, num_proc=32)

    # Make sure that the images are decoded, it prevents from storing bytes.
    # More info here https://github.com/huggingface/blog/pull/2148#discussion_r1667400478
    f = dataset.features
    f["images"] = features.Sequence(features.Image(decode=True))
    dataset = dataset.cast(f)

    # Train the model
    training_args = DPOConfig(
        output_dir="idefics2-8b-dpo",
        bf16=True,
        gradient_checkpointing=True,
        per_device_train_batch_size=2,
        gradient_accumulation_steps=32,
        num_train_epochs=1,

>         dataset_num_proc=32,

Reviewer: we already mapped, I was wondering why we pass parallel proc args

I think they make sense when one passes data collator with batching inside with transformers Trainer, I'm not super familiar with TRL abstractions tbh so I might be wrong

Author: `DPOTrainer` tokenizes internally with `dataset.map`, that's why we use `dataset_num_proc`

| VCD              | 71.8     | 74.9 |
| Less-is-more     | 72.4     | 75.8 |
| OPERA            | 75.2     | 78.3 |
| LURE             | 73.5     | 77.7 |
| QWEN-VL          | 81.9     | 86.4 |
| LLaVA-NeXT       | 81.4     | 85.4 |
| MiniGemini       | 82.6     | 87.6 |
| GPT-4o           | 84.5     | 85.3 |
| Idefics2         | 85.8     | 89.1 |
| **Idefics2+DPO** | 85.9     | 89.4 |

Overall, the model seems to hallucinate a bit less. The training seems to have been successful!

As an illustration, here are some examples where the model hallucinates less:

| Image | Question | Idefics2 | Idefics2+DPO |
| ----- | -------- | -------- | ------------ |
| ![](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/dpo_vlm/AMBER_2.jpg) | Are there two ships in this image? | Yes | No |
| ![](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/dpo_vlm/AMBER_111.jpg) | Is the ground uneven in this image? | No | Yes |
| ![](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/dpo_vlm/AMBER_7.jpg) | Is there one shovel in this image? | Yes | No |

### Finetuning Llava 1.5, PaliGemma and others

> At the time of writing, the DPO implementation in TRL supports Idefics2, Llava 1.5 and PaliGemma, and we're working on adding support for more models. To finetune these models the easier is to use the [example script](https://github.com/huggingface/trl/blob/main/examples/scripts/dpo_visual.py) provided in the TRL repository. For example, to finetune PaliGemma, you can use the following command:

Reviewer: ```suggestion
At the time of writing, the DPO implementation in TRL supports Idefics2, Llava 1.5 and PaliGemma, and we're working on adding support for more models. The easiest way to fine-tune these models is to use the [example script](https://github.com/huggingface/trl/blob/main/examples/scripts/dpo_visual.py) provided in the TRL repository. For example, to finetune PaliGemma, you can use the following command:
```

```sh
accelerate launch examples/scripts/dpo_visual.py \
    --dataset_name HuggingFaceH4/rlaif-v_formatted \
    --model_name_or_path google/paligemma-3b-pt-224 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 32 \
    --dataset_num_proc 32 \
    --output_dir dpo_paligemma_rlaif-v \
    --bf16 \
    --torch_dtype bfloat16 \
    --gradient_checkpointing \
    --use_peft \
    --lora_target_modules=all-linear
```

> You can also refer to the [smol-vision](https://github.com/merveenoyan/smol-vision) project where you'll find more examples and scripts to finetune vision models with DPO.

Reviewer: We should end on an enthusiastic tone, call people to train their own DPO VLMs

response_text = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
print(response_text)
```

As mentioned above, the [openbmb/RLAIF-V-Dataset](https://huggingface.co/datasets/openbmb/RLAIF-V-Dataset) is designed to reduce hallucinations. But has the fine-tuning actually reduced hallucinations? To find out, we can use the [AMBER benchmark](https://arxiv.org/abs/2311.07397), a dataset specifically created to evaluate hallucinations in VLMs. We will report the results for Idefics2 and Idefics2+DPO on the discriminative task and compare them with other models for reference.

|                  | Accuracy | F1       |
| ---------------- | -------- | -------- |
| GPT-4o           | 88.8     | 91.6     |
| **Idefics2+DPO** | **85.9** | **89.4** |
| Idefics2         | 85.8     | 89.1     |
| GPT-4v           | 83.4     | 87.4     |
| MiniGemini       | 82.6     | 87.6     |
| LLaVA-NeXT       | 81.4     | 85.4     |
| QWEN-VL          | 81.9     | 86.4     |
| LURE             | 73.5     | 77.7     |
| OPERA            | 75.2     | 78.3     |
| Less-is-more     | 72.4     | 75.8     |
| VCD              | 71.8     | 74.9     |

> Overall, the fine-tuned model seems to hallucinate a bit less. The training seems to have been successful!

Reviewer: Maybe we should put a link to model repo and demo for people to try

<gradio-app theme_mode="light" space="HuggingFaceH4/compare_idefics-8b-dpo"></gradio-app>

## Finetuning Llava 1.5, PaliGemma and others

At the time of writing, the DPO implementation in TRL supports Idefics2, Llava 1.5, and PaliGemma, with ongoing efforts to add support for more models. The easiest way to fine-tune these models is to use the [example script](https://github.com/huggingface/trl/blob/main/examples/scripts/dpo_visual.py) provided in the TRL repository. For example, to finetune PaliGemma, you can use the following command:

```sh
accelerate launch examples/scripts/dpo_visual.py \
    --dataset_name HuggingFaceH4/rlaif-v_formatted \
    --model_name_or_path google/paligemma-3b-pt-224 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 32 \
    --dataset_num_proc 32 \
    --output_dir dpo_paligemma_rlaif-v \
    --bf16 \
    --torch_dtype bfloat16 \
    --gradient_checkpointing \
    --use_peft \
    --lora_target_modules=all-linear
```

You can find a detailed focus on PaliGemma finetuning in the [smol-vision](https://github.com/merveenoyan/smol-vision) project.

> 🚀🚀 Now you have everything you need to start fine-tuning your own VLMs with DPO. Share your findings, models, and datasets with the community!

Reviewer 2: ```suggestion
🚀🚀 Now you have everything you need to start fine-tuning your VLMs with DPO. Share your findings, models, and datasets with the community! As an extra challenge, we would also encourage you to implement the [mDPO](https://arxiv.org/abs/2406.11839v1) method that augments the standard DPO loss with two auxiliary losses.
```

Reviewer: I honestly think this is a bit overkill :D

dataset = dataset.cast(f)
```

Our dataset is now formatted. Let's have a look at the first example:

```python
>>> dataset[1]
{'images': [<PIL.JpegImagePlugin.JpegImageFile image mode=L size=980x812 at 0x154505570>],
 'prompt': 'User:<image>how many families?<end_of_utterance>\n',
 'rejected': 'Assistant: The image does not provide any information about families.<end_of_utterance>\n',
 'chosen': 'Assistant: The image shows a Union Organization table setup with 18,000 families.<end_of_utterance>\n'}
```

Warm up your GPUs, the dataset is ready for training!

## Training

For the sake of the example, we'll be training the [Idefics2-8b](https://huggingface.co/HuggingFaceM4/idefics2-8b) model, but note that the DPO implementation in TRL supports other models like [Llava 1.5](https://huggingface.co/llava-hf/llava-1.5-7b-hf) and [PaliGemma](https://huggingface.co/google/paligemma-3b-pt-224). More information in Section [Finetuning Llava 1.5, PaliGemma and others](#finetuning-llava-15-paligemma-and-others). Before looking into the training process, we'll first ensure everything fits smoothly into memory.

### How much memory do I need?

> I myself have a GPU with 80GB of VRAM. Is it enough to train my Idefics2-8b model? Here are the calculations steps to get a rough estimate of the memory needed.

Reviewer 2: ```suggestion
I have a GPU with 80GB of VRAM. Is it enough to train my Idefics2-8b model? Here are the calculation steps to get a rough estimate of the memory needed.
```

> Let \( N \) be the number of parameters, \( P \) the precision. The following components require will have to fit together in memory:

Reviewer 2: ```suggestion
Let \( N \) be the number of parameters, \( P \) the precision. The following components will have to fit together in memory:
```

To implement `bfloat16` precision in the model:

```python
import torch
from transformers import AutoModelForVision2Seq

model = AutoModelForVision2Seq.from_pretrained("HuggingFaceM4/idefics2-8b", torch_dtype=torch.bfloat16)
```

`bfloat16` precision can also be applied to the optimizer by setting `bf16=True` in the training arguments:

```python
from transformers import TrainingArguments

training_args = TrainingArguments(..., bf16=True)
```

### LoRA

[LoRA](https://arxiv.org/abs/2106.09685) is a method that reduces the number of trainable parameters by learning pairs of rank-decomposition matrices while keeping the original weights frozen. This significantly decreases the storage needs for LLM adapted to specific tasks. LoRA is integrated in [PEFT](https://github.com/huggingface/peft) and you can set it up in no time:

```diff
  from transformers import AutoModelForVision2Seq
+ from peft import get_peft_model, LoraConfig

  model = AutoModelForVision2Seq.from_pretrained("HuggingFaceM4/idefics2-8b")
+ peft_config = LoraConfig(target_modules="all-linear")
+ model = get_peft_model(model, peft_config)
```

> PEFT acts like a wrapper (called _adaptater_) around the model. This is this adapter that will be trained while the inner model is kept frozen. How much LoRA reduces the number of trainable parameters?

Reviewer 2: ```suggestion
PEFT acts like a wrapper (called _adaptater_) around the model. This is the adapter that will be trained while the inner model is kept frozen. How much does LoRA reduce the number of trainable parameters?
```

Try it yourself and see how the model performs on your own examples!

<script type="module" src="https://gradio.s3-us-west-2.amazonaws.com/4.36.1/gradio.js"></script>

<gradio-app theme_mode="light" space="HuggingFaceH4/compare_idefics-8b-dpo"></gradio-app>

## Finetuning Llava 1.5, PaliGemma and others

At the time of writing, the DPO implementation in TRL supports Idefics2, Llava 1.5, and PaliGemma, with ongoing efforts to add support for more models. The easiest way to fine-tune these models is to use the [example script](https://github.com/huggingface/trl/blob/main/examples/scripts/dpo_visual.py) provided in the TRL repository. For example, to finetune PaliGemma, you can use the following command:

```sh
accelerate launch examples/scripts/dpo_visual.py \
    --dataset_name HuggingFaceH4/rlaif-v_formatted \
    --model_name_or_path google/paligemma-3b-pt-224 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 32 \
    --dataset_num_proc 32 \
    --output_dir dpo_paligemma_rlaif-v \
    --bf16 \
    --torch_dtype bfloat16 \
    --gradient_checkpointing \
    --use_peft \
    --lora_target_modules=all-linear
```

> You can find a detailled focus on PaliGemma finetuning in the [smol-vision](https://github.com/merveenoyan/smol-vision) project.

Reviewer 2: ```suggestion
You can find a detailed focus on PaliGemma finetuning in the [smol-vision](https://github.com/merveenoyan/smol-vision) project.
```

> 🚀🚀 Now you have everything you need to start fine-tuning your own VLMs with DPO. Share your findings, models and datasets with the community!

Reviewer 2: ```suggestion
🚀🚀 Now you have everything you need to start fine-tuning your own VLMs with DPO. Share your findings, models, and datasets with the community!
```

Author: Oxford comma team ;)

thumbnail: /blog/assets/dpo_vlm/thumbnail.png
authors:
- user: qgallouedec
- user: vwxyzjn
- user: merve
---

# Direct Preference Optimization for Visual Language Models with TRL

Training models to understand and predict human preferences can be incredibly complex. Traditional methods, like supervised fine-tuning, often require assigning specific labels to data, which can be difficult, especially for nuanced tasks. There is an alternative approach that can simplify this process and yield more accurate results: preference optimization. By focusing on comparing and ranking candidate answers rather than assigning fixed labels, preference optimization allows models to capture the subtleties of human judgment more effectively.

Preference optimization is widely used for finetuning language models, but it can also be applied to visual language models (VLM).
We are excited to announce that the [TRL](https://huggingface.co/docs/trl/index) library now supports direct preference optimization for VLMs. This article will guide you through the process of training VLMs using TRL and direct preference optimization.

## Dataset

> Preference optimisation requires data that captures user preferences. For example, you need to have samples like the following:

Reviewer: ```suggestion
Preference optimization requires data that captures user preferences. For example, you need to have samples like the following:
```

+ peft_config = get_peft_config(model_config)
+ model = get_peft_model(model, peft_config)
```

How much LoRA reduces the number of trainable parameters?

```python
>>> model.print_trainable_parameters()
trainable params: 55,348,736 || all params: 8,458,116,848 || trainable%: 0.6543860411799315
```

It reduces the number of trainable parameters from 8 billion to 55 million, which is a huge gap.

Now that we have reduced the memory requirements, let's recalculate the memory needed:

| Component          | Calculation                           | Memory (GB) |
| ------------------ | ------------------------------------- | ----------- |
| Model to train     | \( 8 \mathrm{G} \times 2 \)           | 16          |
| Reference model    | \( 8 \mathrm{G} \times 2 \)           | 16          |
| Gradients          | \( 55 \mathrm{M} \times 2 \)          | 0.2         |
| Optimizer states   | \( 2 \times 55 \mathrm{M} \times 2 \) | 0.4         |
| **Total**          |                                       | **32.6**    |

This time, we need around 30GB of memory to finetune our Idefics2, which is much more reasonable and fits within my GPU!

> 

Author: Here we use just Lora (not QLoRA). Does it still add a significant overhead?

#### What about the batch size?

> Our memory calculation isn't exact as it doesn't account for activations, which are the intermediate outputs of the network layers. Activations depend on the batch size, which affects the overall memory usage. To choose an appropriate training batch size, a practical approach is to start with your desired batch size (e.g., 64). You will likely encounter an OOM error with this size. If so, halve the batch size until the memory fits within your GPU. For each reduction by half, double the gradient accumulation steps to maintain the same effective batch size. In our case, we end up with a batch size of 2 and gradient accumulation steps of 32.

Reviewer: actually, google has a good guide here for LLMs https://cloud.google.com/vertex-ai/generative-ai/docs/model-garden/lora-qlora indeed it doesn't account for activations but nearly all LLMs have same ops these days

Author: 6c15dde

    )

    trainer.train()

if __name__ == "__main__":
    main()
```

Let's run and wait... 🚀

```sh
accelerate launch dpo_idefics2-8b.py
```

### Results

A few hours later, the training is complete. Let's have a look at the training curves:

![Learning curves](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/dpo_vlm/learning_curves.png)

In DPO, we focus on some metrics to determine if the training went well.

- **The accuracy**: the percentage of training samples for which the model has a higher chance of outputting the chosen answer than the rejected answer. We can see that the accuracy is increasing, which is a good sign.
- **The rewards**: the reward are somewhat related to the probability of an answer being chosen. We refer the reader to the TODO for more details. We expect the reward of the chosen answer to be higher than the rejected answer. To check this, we can watch the _reward margin_ which is the difference between the rewards of the chosen answer and the rejected answer. We can see that the reward margin is increasing, which is a also good sign.

--------------------------------------------

> - Les paramètres du modèle à entrainer

Reviewer: pourquoi changer à Français? 😅

Author: Everything below 289 is my draft 😅

Dans notre cas, le calcul est le suivant:

| Composant           |                                             | Calcul                                 | Mémoire en Gb |
| ------------------- | ------------------------------------------- | -------------------------------------- | ------------- |
| Modèle à entrainer  | Nombre de paramètres \( \times \) Precision | 8b \( \times \) 4 bytes (float32)      | 32            |
| Modèle de référence | Même que le modèle à entrainer              | 8b \( \times \) 4 bytes                | 32            |
| Gradients           | Même que le modèle à entrainer              | 8b \( \times \) 4 bytes                | 32            |
| Optimiseur          | 2* modèle à entrainer (AdamW)               | 2 \( \times \) 8B \( \times \) 4 bytes | 64            |
| **Total**           |                                             |                                        | **160**       |

Aïe, 160 est bien au deussus de ce que mon hardware peut contenir (80Gb). Il va falloir réduire la quantité de mémoire utilisée.

> Parameters efficient training

Reviewer: ```suggestion
### Parameter Efficient Fine-tuning
```

> Work has shown that it is possible to reduce the search set to a small set of parameters. In particular, LoRA  is a method that reduces the number of trainable parameters by learning pairs of rank-decomposition matrices while keeping the original weights frozen. This significantly decreases the storage needs for large language models adapted to specific tasks.

Reviewer: everything below to be removed later
